### DISCUSSION TITLE
Understanding space accounting and recordsize

### DISCUSSION THREAD

**Discussion by frukto** [openzfs/zfs#13055] [General]
I have 8 disks raidz2 pool with `ashift=13`. I created a bunch of datasets with different `recordsize` and did write 64M of data parted in files of various sizes:
 - 65536 1K files
 - 32768 2K files
 - ...
 - 2 32M files

I checked the `used` property of the file system and captured the values in the table below.

For any combination of `recordsize` and the patterns of `number_of_files` and `file_size` the test was:

```
zfs create pool/fs -o recordsize=${recordsize} -o mountpoint=/mnt/fs
for n  in $(seq 1 ${number_of_files}) ; do
    dd if=/dev/urandom of=/mnt/fs/$n bs=${file_size} count=1
done
# store the value of: zfs get used pool/fs
zfs destroy pool/fs
```
I am having trouble to make sense of the values. Why is there such a high overhead for writing (small) files, in particular for small `recordsize`?

recordsize | 65536@1K | 32768@2K | 16384@4K | 8192@8K | 4096@16K | 2048@32K | 1024@64K | 512@128K 
-----------|----------|----------|----------|---------|----------|----------|----|---
4K | 1103.85M | 552.27M | 276.44M | 522.23M | 389.30M | 322.68M | 289.21M | 272.76M
8K | 1103.79M | 552.30M | 276.44M | 138.55M | 261.49M | 194.84M | 161.40M | 144.95M 
16K | 1103.82M | 552.30M | 276.35M | 138.58M | 133.61M | 194.93M | 161.68M | 144.88M 
32K | 1103.82M | 552.24M | 276.41M | 138.52M | 133.55M | 67.02M | 97.49M | 81.10M 
64K | 1103.85M | 552.36M | 276.54M | 138.61M | 133.52M | 67.09M | 65.46M | 81.04M
128K | 1103.88M | 552.27M | 276.44M | 138.58M | 133.55M | 67.02M | 65.46M | 64.93M 
256K | 1103.82M | 552.24M | 276.47M | 138.58M | 133.61M | 67.21M | 65.59M | 64.96M 

recordsize |  256@256K  | 128@512K | 64@1M | 32@2M | 16@4M | 8@8M | 4@16M | 2@32M
------|----------|-----------|----------|-------|-------|-------|------|-----
4K | 264.49M | 260.37M | 260.30M | 258.37M | 257.74M | 257.96M | 257.81M | 257.71M
8K | 136.86M | 132.52M | 130.40M | 130.40M | 129.40M | 129.12M | 129.21M | 129.09M
16K | 136.64M | 132.71M | 130.43M | 129.46M | 129.34M | 128.87M | 128.68M | 128.71M
32K | 72.74M | 68.58M | 66.52M | 65.56M | 64.96M | 64.93M | 64.65M | 64.53M
64K | 72.74M | 68.58M | 66.52M | 65.49M | 64.93M | 64.68M | 64.65M | 64.46M
128K | 72.67M | 68.58M | 66.49M | 65.49M | 64.93M | 64.68M | 64.49M | 64.46M
256K | 60.87M | 64.62M | 62.50M | 61.43M | 60.97M | 60.72M | 60.50M | 60.40M

**Comment by gmelikov** [comment]
- you can't write a block in less than 2^ashift bytes sector size (embedded_block feature aside), sector size is a minimal block size to write
- in raidz you have to have a padding block per every block, which can't be less than minimal sector size too

Useful table https://docs.google.com/spreadsheets/d/1tf4qx1aMJp8Lo_R6gpT689wTjHv6CGVElrPqTA0w_ZY/edit?pli=1#gid=2126998674

**Comment by gmelikov** [comment]
I've created the issue to document it better https://github.com/openzfs/openzfs-docs/issues/263

**Reply by frukto** [reply]
With `ashift=13` the minimal write is 8K. So for example 65536 writes à 1K result into 65536 blocks of 8K (that is 512M). I did not understand your second point though. Why is there a padding block for every block?
Did you mean parity block? Even then, I would expect two parity blocks per data block. Something like this, for the first three data blocks, that is for each file 8K data plus 2 x 8K parity, resulting in even more space consumption (~1.5G)
```
  0      1      2      3      4      5      6       7
[ P ]  [ P ]  [ D ]  [ P ]  [ P ]  [ D ]  [ P ]  [ P ]
[ D ] ...
```
Thanks for opening the issue, a clarifying documentation would be very helpful.

**Reply by rincebrain** [reply]
[The blogpost that goes with that table is informative as well.](https://www.delphix.com/blog/delphix-engineering/zfs-raidz-stripe-width-or-how-i-learned-stop-worrying-and-love-raidz)

**Reply by frukto** [reply]
Let's assume
- `B`: sectors to write (rows in the spreadsheet)
- `P`: raidz level, e.g. P=2 for raidz2 
- `N`: number of disks (columns in the spreadsheet)
- `D`: N - P

We can write up to `D` sectors before we have to add `P` parity sectors, afterwards we can write again up to `D` sectors before adding parity information again, .. All in all we need to add `floor(B/D + (D-1)/D)` times the parity information and each time `P` sectors to be precise.

So far we would write `B + P * floor(B/D + (D-1)/D)` sectors (of data and parity), but also need to pad the whole thing up to be a multiple of `P+1` sectors, so 

```
sectors to write =  round_to_next_multiple( B + P * floor(B/D + (D-1)/D) ; P+1)
```
This agrees with the calculation in the spreadsheet. However this does not (even roughly) agree with the measured values of my table above.

**Reply by rincebrain** [reply]
You may find rooting around with zdb on the individual files to be a faster way to explore how much gets allocated for a given example, and what kind of blocks it's spent on.

My anticipation would be that as your recordsize gets tinier, you're spending a higher and higher percentage of used space on the metadata overhead for tracking the data blocks.

**Reply by frukto** [reply]
Yes I am playing around with `zdb`. At the moment I am looking at `recordsize=16K`. I write out 65536 1K files (numbered from 1 to 65536).

`zdb` shows the objects all similar to: 
```
    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
        73    1   128K     1K    16K     512     1K  100.00  ZFS plain file (K=inherit) (Z=inherit=lz4)
                                               176   bonus  System attributes
	dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED 
	dnode maxblkid: 0
	path	/596
	uid     0
	gid     0
	atime	Thu Feb  3 15:53:23 2022
	mtime	Thu Feb  3 15:53:23 2022
	ctime	Thu Feb  3 15:53:23 2022
	crtime	Thu Feb  3 15:53:23 2022
	gen	2864331
	mode	100644
	size	1024
	parent	34
	links	1
	pflags	840800000004
Indirect blocks:
               0 L0 0:55f6193e000:6000 400L/400P F=1 B=2864331/2864331 cksum=88a761a34a:45f85b680812:17d6720d6d20a8:6159de6b9d43fad

		segment [0000000000000000, 0000000000000400) size    1K
```
I am trying to read/interpret this information. 
```
Indirect blocks:
               0 L0 0:55f6193e000:6000 400L/400P F=1 B=2864331/2864331 cksum=...
```
- `L0` this is a "level 0" direct block ??
- written at offset `0x55f6193e000`  (with size `0x6000` = 24K = 3 * 2^`ashift` ??)
- logical = physical size = `0x400` = 1K
- `F=1` ??
- written in transaction group `2864331` ??

Comparing the offsets, I see the files having a "distance" of 24K (data + 2 x parity ?). All the files sum up to a change of dataset's `used` by  ~1103M (= 65536 * 17.2KB) . No matter how I do the math things don't line up.

**Reply by rincebrain** [reply]
Ah, you're looking at the value of `zfs get used`, I overlooked that in your initial post. You may find it more informative to look at `zpool get alloc,free` before and after, as well, and compare the two, probably with -p to get exact values.

**Reply by JBigglesworth** [reply]
I'm coming in here somewhat naively and I'd welcome any correction if my following input is wrong or misguided.

My crude understanding is that you're talking in term of numbers of sectors but, even if you're only writing just one byte of data you're incurring a write of an 8KB data block (ashift=13) plus one 8KB block of parity (RaidiZ1) before you go anywhere. (ie: 1 byte data => 16KB disk space).

It leads me to think that block size becomes problematic for RAIDZx as target file size reduces and also that recordsize is immaterial.

Putting aside redundacy and write/read performance concerns, wouldn't this mean that for small files sizes, where block size seems to be more significant that 2 way mirrors are more space efficient than RAIDZ for small file sizes?

**Reply by rincebrain** [reply]
raidz and small records are not the ideal mix, which is one reason you can do things like shove {metadata, small records} onto a {mirrored} special vdev.

**Reply by JBigglesworth** [reply]
@rincebrain
Thanks for mentioning special vdevs, I'd heard of them but haven't experimented with them yet.  They seem a much better alternative than partioning data via multiple datasets to optimise data storage!

**Reply by frukto** [reply]
@JBigglesworth 
>  a write of an 8KB data block (ashift=13) plus one 8KB block of parity (RaidiZ1) before you go anywhere. (ie: 1 byte data => 16KB disk space).

Exactly, even worse with raidz2, that would be 24K. That is how I am understanding this as well. 

Maybe the confusion is just about `zfs get used` vs `zpool get alloc`? Comparing values of the latter, I do see roughly 1.5G used for 65536 1K files. That is sound. But `zfs get used` shows me 1103.85M.

My understanding of the man page is that `allocated` does include everything (parity, padding, metadata,...) but what about `used` then?

Reading about "raidz deflation" in [lib/libzfs/libzfs_dataset.c#L5285](https://github.com/openzfs/zfs/blob/399b98198ae2e7c003448bf6b9508879c2a3ed83/lib/libzfs/libzfs_dataset.c#L5285):

In an 8 disk wide raidz2 with `ashift=13` a 128K block would be written as `192K = 24 x 8K = 16 x 8K data + 6 x 8K parity + 2 x 8K skip blocks`, so `raidz_deflation = 128K / 192K = 2/3`. Is `used` just scaled by this factor. So the real 1.5G are accounted as `2/3 * ~1.5G = 1103.85M`? That would at least roughly match. 

Also everything above does not consider any metadata, so probably its even more complicated.
